docs/react/guides/advanced-ssr.md

```diff
@@ -0,0 +1,390 @@
+      }),
+  )
+
+  return (
+    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
+  )
+}
+```
+
+```tsx
+// In Next.js, this file would be called: app/layout.jsx
+import Providers from './providers'
+
+export default function RootLayout({ children }) {
+  return (
+    <html lang="en">
+      <head />
+      <body>
+        <Providers>{children}</Providers>
+      </body>
+    </html>
+  )
+}
+```
+
+This part is pretty similar to what we did in the SSR guide, we just need to split things up into two different files.
+
+### Prefetching and de/hydrating data
+
+Let's next look at how to actually prefetch data and dehydrate and hydrate it. This is what it looked like using the **Next.js pages router**:
+
+```tsx
+// pages/posts.jsx
+import { dehydrate, HydrationBoundary, QueryClient, useQuery } from '@tanstack/react-query'
+
+// This could also be getServerSideProps
+export async function getStaticProps() {
+  const queryClient = new QueryClient()
+
+  await queryClient.prefetchQuery({
+    queryKey: ['posts'],
+    queryFn: getPosts,
+  })
+
+  return {
+    props: {
+      dehydratedState: dehydrate(queryClient),
+    },
+  }
+}
+
+function Posts() {
+  // This useQuery could just as well happen in some deeper child to
+  // the <PostsRoute>, data will be available immediately either way
+  const { data } = useQuery({ queryKey: ['posts'], queryFn: getPosts })
+
+  // This query was not prefetched on the server and will not start
+  // fetching until on the client, both patterns are fine to mix
+  const { data: commentsData } = useQuery({
+    queryKey: ['posts-comments'],
+    queryFn: getComments,
+  })
+
+  // ...
+}
+
+export default PostsRoute({ dehydratedState }) {
+  return (
+    <HydrationBoundary state={dehydratedState}>
+      <Posts />
+    </HydrationBoundary>
+  )
+}
+```
+
+Converting this to the app router actually looks pretty similar, we just need to move things around a bit. First, we'll create a Server Component to do the prefetching part:
+
+```tsx
+// app/posts/page.jsx
+import {
+  dehydrate,
+  HydrationBoundary,
+  QueryClient,
+} from '@tanstack/react-query'
+import Posts from './posts'
+
+export default async function PostsPage() {
+  const queryClient = new QueryClient()
+
+  await queryClient.prefetchQuery({
+    queryKey: ['posts'],
+    queryFn: getPosts,
+  })
+
+  return (
+    // Neat! Serialization is now as easy as passing props.
+    // HydrationBoundary is a Client Component, so hydration will happen there.
+    <HydrationBoundary state={dehydrate(queryClient)}>
+      <Posts />
+    </HydrationBoundary>
+  )
+}
+```
+
+Next, we'll look at what the Client Component part looks like:
+
+```tsx
+// app/posts/posts.jsx
+'use client'
+
+export default function Posts() {
+  // This useSuspenseQuery could just as well happen in some deeper
+  // child to <Posts>, data will be available immediately either way
+  const { data } = useSuspenseQuery({ queryKey: ['posts'], queryFn: getPosts })
```

Comment:
does this really work? I thought if we `useSuspenseQuery` in the app directory, it will enable streaming 🤔

Reply:
If we have already awaited the data so it's in the cache and it's not stale, we shouldn't suspend so it should work? But I do think we need to add `staleTime` here and point that out.

As an aside we should probably have a full codesandbox example with this pattern (prefetching+suspense), either for the pages router, the app router or both.

Reply:
Hmm, yeah, it's a problem. It's like if this is a weird middle ground between `useQuery` and `useSuspenseQuery`, where neither fits perfectly. It _should_ never suspend, it should always have data, it should never have the loading state. I'll think some more but maybe it is best to just do `useQuery` here for now.

Reply:
I just went ahead and changed it out to `useQuery` and did some tweaks in other places to reflect this. Can revisit later if we want to but I agree `useQuery` probably makes the most sense right now.